<a href="https://colab.research.google.com/github/LMORALESDEV/PYTHON-ITLA/blob/main/0302_Funciones.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Temas a tratar

1. Cómo definir una función en Python
2. Cómo usar o llamar a una función
3. Sentencia RETURN
4. Argumentos indeterminados o arbitrarios
5. Funciones recursivas
6. Funciones Integradas

# Funciones en Python

*   Una función es un bloque de código que solo se ejecuta cuando se llama o invoca.
*   Se puede pasar datos, conocidos como parámetros, a una función.

*   Una función puede devolver datos como resultado.

#Crear una función

Para crear una función se utiliza la palabra reservada ***def*** seguido del nombre de la función. Luego los paréntesis, dos puntos y el cuerpo de la función que se coloca en las siguientes líneas. Si la función recibe argumentos se colocan dentro de los paréntesis.
Para ejecutar la función hay que invocarla. Para invocar la función se coloca el nombre de la función seguido de los paréntisis.

In [ ]:
def hola():                               #Nombre de la función
  print("Hola desde la función hola")     #Cuerpo de la función. Se ejecuta la función print

hola()                                    #Se invoca a la función hola().

# Parámentro versus Argumento

*   Los términos parámetro y argumento pueden usarse para lo mismo: información que se pasa a una función.
*   Desde la perspectiva de una función:
    *   Un parámetro es la variable listada entre paréntesis en la definición de la función.
    *   Un argumento es el valor que se envía a la función cuando se
llama o invoca.

#Argumentos

La información que se le pasa a las funciones se les llama argumentos. Los argumentos se definen dentro de los paréntesis y se separan por comas.

La función *sumar* tiene definido dos parámetros: *a* y *b*. Cuando se llama a la función, se le pasa los argumentos 4 y 5. La función hace la operación suma e imprime el resultado igual a 9. Luego se vuelve a llamar la función *sumar* con los argumentos 7 y 3 y el resultado es 10.

In [ ]:
def sumar(a, b):
  resultado = a + b
  print(resultado)

#Invocar la función
sumar(4, 5)
sumar(7, 3)

9
10


#Argumentos arbitrarios o indeterminados: POR POSICION

Si no se sabe cuántos argumentos se pasarán a su función, se puede agregar un asterisco (*) antes del nombre del parámetro en la definición de la función. Los parámetros se recibirán en forma de una tupla y se pueden indexar por su posición en la tupla inciando en 0.

En este ejemplo la función *mi_funcion* utiliza asterisco antes del parámetro *personas* para recibir los argumentos. Estos serán recibidos en forma de tupla. Con la función *print* se imprime la tupla y luego con la función *print* se imprime la segunda posición de la tupla.



In [ ]:
#EJEMPLO 1: Imprimir un valor segun su indice
def mi_funcion(*personas):
  print(personas)
  print("El nombre de la doctora es: " + personas[1])

#Invocar la función
mi_funcion("Luis", "María", "Carlos", "Jean", "Dilan","Pedro")


('Luis', 'María', 'Carlos', 'Jean', 'Dilan', 'Pedro')
El nombre de la doctora es: María


In [ ]:
#EJEMPLO 2: suma de varios numeros
def sumar(*numeros):
    resultado = 0
    for num in numeros:
        resultado += num
    return resultado

print(sumar(1, 2, 3))
print(sumar(4, 5, 6, 7,10,12,11,15,20,21,32,100))


6
243


# Argumentos usando palabras clave

También se puede enviar argumentos con la sintaxis *clave = valor*. De esta manera, el orden de los argumentos no importa.

En este ejemplo, se le envía los argumentos a la función usando la sintaxis *clave = valor* y aunque se envién en diferente orden, no se afecta el valor que le corresponde a cada argumento.

In [ ]:
def mi_funcion(hijo3, hijo2, hijo1):
  print("El hijo3 es " + hijo3)

#Invocar la función
#mi_funcion("Rosa", "Ana", "María")
mi_funcion(hijo2 = "Rosa", hijo1 = "Ana", hijo3 = "María")

El hijo3 es María


#Argumentos arbitrarios o indeterminados: POR NOMBRE


Si no se sabe cuántos argumentos de palabras clave se pasarán a su función, se agrega dos asteriscos: ** antes del nombre del parámetro en la definición de la función.

En este ejemplo, la función recibirá un diccionario de argumentos y podrá acceder a los elementos que se desea indexando el diccionario por la clave.

In [ ]:
#EJEMPLO 1: Recibiendo un diccionario
def mi_funcion(**persona):
  print("Su sector es " + persona["dirreccion"])

#Invocar la función
mi_funcion(nombre = "Ana", apellido = "Rojas", edad = 32, pais = "RD", sector="San Isidro")

Su sector es Calle 32


In [ ]:
def imprimir_parametros(**parametros):
    for clave, valor in parametros.items():
        print(clave + " = " + str(valor))

imprimir_parametros(nombre="Juan", edad=30, ciudad="Madrid", puesto = "Asesor", idioma = "ES", genero = "M", carrera = "Ing")


nombre = Juan
edad = 30
ciudad = Madrid
puesto = Asesor
idioma = ES
genero = M
carrera = Ing


##Funcion que acepte argumentos indeterminados por POSICION y por NOMBRE

En este ejemplo, la función calcular_precio() toma un argumento obligatorio cantidad, seguido de un número variable de descuentos opcionales y un número variable de impuestos con palabras clave. Los descuentos se restan de la cantidad, y los impuestos se agregan multiplicando la cantidad por su tasa correspondiente. El resultado final es el precio total. En este ejemplo, se le pasa una cantidad de 100, dos descuentos de 10 y 5, y dos impuestos de 18% e 10%. El precio total resultante es 126.

In [ ]:
def calcular_precio(cantidad, *descuentos, **impuestos):
    precio_total = cantidad
    for descuento in descuentos:
        precio_total -= descuento
    for clave, valor in impuestos.items():
        precio_total += cantidad * valor
    return precio_total

total_bruto = float(input("Ingrese el subtotal"))
d1 = int(input("1er Descuento"))
d2 = int(input("2do Descuento"))
d3 = int(input("3er Descuento"))
#print(calcular_precio(100, 10 , 7, 5,itbis=0.18, tasa_servicio=0.05, envio = 0.10, comision = 0.03 ))
print(calcular_precio(total_bruto, d1, d2, d3 ))
#print(calcular_precio(total_bruto, imp_global = .33 ))


Ingrese el subtotal150
1er Descuento10
2do Descuento15
3er Descuento5
120.0


#Valor de parámetro predeterminado

Un valor de parámetro predeterminado se utiliza para cuando no se le envía argumentos a una función. Entonces la función utiliza el parámetro predeterminado.

En este ejemplo, Si llamamos a la función sin argumento, utiliza el valor predeterminado que es *Costa Rica*.

In [ ]:
def mi_funcion(país = "Costa Rica" ):
  print("Yo soy de  " + país)

#Invocar a la función
mi_funcion("Nicaragua")
mi_funcion("España")
mi_funcion()
mi_funcion("Brazil")

Yo soy de  Nicaragua
Yo soy de  España


TypeError: mi_funcion() missing 1 required positional argument: 'país'

#Pasar una lista como argumento

Se puede enviar cualquier tipo de argumento de datos a una función (cadena, número, lista, diccionario, etc.), y se tratará como el mismo tipo de datos dentro de la función.

En este ejemplo, se le envía una lista como parámetro a la función.

In [ ]:
def mi_funcion(personas):
  print(type(personas))
  for x in personas:
    print(x)

#Invocar a la función
personas = ["Ana", "Carlos", "Rosa"]
mi_funcion(personas)

<class 'list'>
Ana
Carlos
Rosa


#Valores de retorno

Para permitir que una función devuelva un valor, se usa la instrucción *return*.

En este ejemplo, la función retorna el valor de la variable resultado al lugar donde fue invocada la función. Ese valor que retorna la función puede ser almacenado en una variable y ser utilizado para otras cálculos.

In [ ]:
def mi_funcion(a, b):
  resultado = a + b
  return resultado

#Invocar a la función
respuesta = mi_funcion(4, 4)
print(respuesta)

respuesta = mi_funcion(9, 5)
respuesta = respuesta + 3
print(respuesta)

8
17


#La declaración pass

 Las definiciones de funciones no pueden estar vacías, pero si por alguna razón se tiene una definición de función sin contenido, se ingresa la instrucción  *pass* para evitar un error.

In [ ]:
# Esta definición vacía genera un error.
def mi_funcion():

SyntaxError: incomplete input (<ipython-input-22-dffc93435264>, line 2)

In [ ]:
def mi_funcion():
    pass

# Funciones Integradas

Las funciones integradas (también conocidas como funciones incorporadas o funciones predefinidas) son funciones que vienen incluidas en el lenguaje Python y se pueden usar directamente en cualquier programa sin necesidad de definirlas previamente.


**print(), input(), type(), int(), str(),float(), etc...** son algunas de las mas conocidas.


A continuación un listado de las funciones integradas no tan comunes.



**range()**: Genera una secuencia de números enteros.

In [ ]:
secuencia = range(5)
for num in secuencia:
    print(num)

**abs()**: Devuelve el valor absoluto de un número.

In [ ]:
numero = -5
print(abs(numero))

5


**max(), min()**: Devuelve el valor máximo o mínimo de una secuencia de valores.

In [ ]:
numeros = [1, 5, 3, 7, 2]
print(max(numeros))
print(min(numeros))

7
1


**enumerate()**: Devuelve un objeto enumerado que contiene pares (índice, valor) de una secuencia.

In [ ]:
nombres = ["Juan", "María", "Pedro"]
for x, y in enumerate(nombres):
    print(x, y)

0 Juan
1 María
2 Pedro


**bin()**: Conversión de entero a binario.

In [ ]:
bin(100)

'0b1100100'

**round()**: Redondeo de un flotante a entero, menor de .5 a la baja, mayor o igual a .5 al alza.

In [ ]:
print(round(7.5))
print(round(7.4))

8
7


**eval()**: Evalúa una cadena como una expresión, acepta variables si se han definido anteriormente.


In [ ]:
#Ejemplo 1
eval('8 + 5')

#Ejemplo 2
v = 10
a = int(input())
eval('2+v+a')

15


27

**help()**: Invoca el menú de ayuda del intérprete de Python.

In [ ]:
help()

#Funciones Recursivas

Se trata de funciones que se llaman a sí mismas durante su propia ejecución. Funcionan de forma similar a las **iteraciones**, pero debemos encargarnos de planificar el momento en que dejan de llamarse a sí mismas o tendremos una función rescursiva infinita.

Suele utilizarse para **dividir una tarea en subtareas** más simples de forma que sea más fácil abordar el problema y solucionarlo.

##Ejemplo de Función SIN retorno(RETURN)

Cuenta regresiva hasta cero a partir de un número:


In [ ]:
def cuenta_regresiva(num):
    num -= 1
    if num > 0:
        print(num)
        cuenta_regresiva(num) #Llamado de la función
    else:
        print("FIN DE LA CUENTA REGRESIVA!")

cuenta_regresiva(10)

9
8
7
6
5
4
3
2
1
FIN DE LA CUENTA REGRESIVA!


##Ejemplo de Función CON retorno(RETURN)

##Factorial de un numero

El factorial de un número corresponde al producto de todos los números desde 1 hasta el propio número. Es el ejemplo con retorno más utilizado para mostrar la utilidad de este tipo de funciones:

Ejemplos:

**Factorial de 3!** = 1 x 2 x 3 = 6

**Factorial de 5!** = 1 x 2 x 3 x 4 x 5 = 120



In [ ]:
def factorial(n):
    #Si n es 0, el factorial es 1
    if n == 0:
        return 1
    else:
        # el factorial de n es n multiplicado por el factorial de n-1
        return n * factorial(n-1)

fact = int(input("Que numero desea buscar el factorial -> "))
print(factorial(fact))

Que numero desea buscar el factorial -> 5
120


##Suma de los primeros "n" números naturales:

In [ ]:
def suma(n):
    #Si n es igual 1, la suma de 1 es 1
    if n == 1:
        return 1
    else:
        #La suma de los primeros n números es n más la suma de los primeros n-1 números
        return n + suma(n-1)

sumatoria = int(input("Que numero desea buscar la sumatoria -> "))
print(suma(sumatoria))


Que numero desea buscar la sumatoria -> 10
55


##Cálculo del máximo común divisor (MCD) de dos números usando el algoritmo de Euclides:

"Euclides explica que el máximo común divisor de dos números se puede encontrar dividiendo el número mayor por el número menor. Si la división es exacta, el m.c.d. es el número menor. Si la división no es exacta, entonces se toma el residuo, y se divide tantas veces como haga falta para llegar a una división sin residuo. El m.c.d. es el último número por cuál se puede dividir."

In [ ]:
def mcd(a, b):
    #Si b es 0, el MCD es a
    if b == 0:
        return a
    else:
        #El MCD de a y b es igual al MCD de b y el resto de a dividido por b
        return mcd(b, a % b)

a = int(input("Cual es el menor valor -> "))
b = int(input("Cual es el mayor valor -> "))
print(mcd(a, b))
